drivers/flash/flash_nxp_s32_qspi_nor.c

```diff
@@ -0,0 +1,1104 @@
+{
+	struct nxp_s32_qspi_data *data = dev->data;
+	Qspi_Ip_MemoryConfigType *memory_cfg = get_memory_config(dev);
+	Qspi_Ip_StatusType status;
+	int ret = 0;
+
+	nxp_s32_qspi_lock(dev);
+
+	status = Qspi_Ip_RunCommand(data->instance, memory_cfg->statusConfig.writeEnableSRLut, 0U);
+	if (status != STATUS_QSPI_IP_SUCCESS) {
+		ret = -EIO;
+	}
+
+	nxp_s32_qspi_unlock(dev);
+
+	return ret;
+}
+
+static int nxp_s32_qspi_write_status_register(const struct device *dev,
+					      uint8_t reg_num,
+					      uint8_t val)
+{
+	const struct nxp_s32_qspi_config *config = dev->config;
+	struct nxp_s32_qspi_data *data = dev->data;
+	Qspi_Ip_StatusType status;
+	uint8_t buf[2] = { 0 };
+	uint16_t lut_idx;
+	size_t size;
+	int ret;
+
+	if (reg_num == 1) {
+		/* buf = [val] or [val, SR2] */
+		lut_idx = QSPI_LUT_IDX(QSPI_SEQ_WRSR);
+		size = 1U;
+		buf[0] = val;
+
+		if (config->qer_type == JESD216_DW15_QER_S2B1v1) {
+			/* Writing SR1 clears SR2 */
+			size = 2U;
+			ret = nxp_s32_qspi_read_status_register(dev, 2, &buf[1]);
+			if (ret < 0) {
+				return ret;
+			}
+		}
+	} else if (reg_num == 2) {
+		/* buf = [val] or [SR1, val] */
+		if ((config->qer_type == JESD216_DW15_QER_VAL_S2B1v1) ||
+		    (config->qer_type == JESD216_DW15_QER_VAL_S2B1v4) ||
+		    (config->qer_type == JESD216_DW15_QER_VAL_S2B1v5)) {
+			/* Writing SR2 requires writing SR1 as well */
+			lut_idx = QSPI_LUT_IDX(QSPI_SEQ_WRSR);
+			size = 2U;
+			buf[1] = val;
+			ret = nxp_s32_qspi_read_status_register(dev, 1, &buf[0]);
+			if (ret < 0) {
+				return ret;
+			}
+		} else {
+			lut_idx = QSPI_LUT_IDX(QSPI_SEQ_WRSR2);
+			size = 1U;
+			buf[0] = val;
+		}
+	} else {
+		return -EINVAL;
+	}
+
+	nxp_s32_qspi_lock(dev);
+
+	status = Qspi_Ip_RunWriteCommand(data->instance, lut_idx, 0U, (const uint8_t *)buf,
+					 (uint32_t)size);
+	if (status == STATUS_QSPI_IP_SUCCESS) {
+		/* Wait for the write command to complete */
+		ret = nxp_s32_qspi_wait_until_ready(dev);
+	} else {
+		LOG_ERR("Failed to write to SR%u", reg_num);
+		ret = -EIO;
+	}
+
+	nxp_s32_qspi_unlock(dev);
+
+	return ret;
+}
+
+static int nxp_s32_qspi_set_quad_mode(const struct device *dev, bool enabled)
+{
+	const struct nxp_s32_qspi_config *config = dev->config;
+	uint8_t sr_num;
+	uint8_t sr_val;
+	uint8_t qe_mask;
+	bool qe_state;
+	int ret;
+
+	switch (config->qer_type) {
+	case JESD216_DW15_QER_NONE:
+		/* no QE bit, device detects reads based on opcode */
+		return 0;
+	case JESD216_DW15_QER_S1B6:
+		sr_num = 1U;
+		qe_mask = BIT(6U);
+		break;
+	case JESD216_DW15_QER_S2B7:
+		sr_num = 2U;
+		qe_mask = BIT(7U);
+		break;
+	case JESD216_DW15_QER_S2B1v1:
+		__fallthrough;
+	case JESD216_DW15_QER_S2B1v4:
+		__fallthrough;
+	case JESD216_DW15_QER_S2B1v5:
+		__fallthrough;
+	case JESD216_DW15_QER_S2B1v6:
+		sr_num = 2U;
+		qe_mask = BIT(1U);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	ret = nxp_s32_qspi_read_status_register(dev, sr_num, &sr_val);
+	if (ret < 0) {
+		return ret;
```

Comment:
Wouldn't it be worth to LOG something here, at least as debug, it seems that there are several `return` lines here which can return exactly the same value with no way to figure which line failed.

Reply:
added err logs to `nxp_s32_qspi_read_status_register()` and `nxp_s32_qspi_write_enable()`